streamingvisitors/src/tests/matching_elements_filler/matching_elements_filler_test.cpp

```diff
@@ -0,0 +1,422 @@
+};
+
+MatchingElementsFillerTest::MatchingElementsFillerTest()
+    : ::testing::Test(),
+      _doc_type(),
+      _struct_field_mapper(make_struct_field_mapper()),
+      _field_path_map(make_field_path_map(_doc_type)),
+      _field_searcher_map(make_field_searcher_map()),
+      _index_to_field_ids(make_index_to_field_ids()),
+      _hit_collector(10),
+      _search_result(),
+      _query(),
+      _shared_searcher_buf(std::make_shared<vsm::SearcherBuf>()),
+      _matching_elements_filler(),
+      _matching_elements(),
+      _sdoc()
+{
+    _search_result.addHit(1, "id::test::1", 0.0, nullptr, 0);
+    _sdoc = std::make_unique<StorageDocument>(_doc_type.make_test_doc(), _field_path_map, _field_path_map->size());
+    EXPECT_TRUE(_sdoc->valid());
+    MatchData md(MatchData::params());
+    _hit_collector.addHit(_sdoc.get(), 1, md, 0.0, nullptr, 0);
+}
+
+MatchingElementsFillerTest::~MatchingElementsFillerTest() = default;
+
+void
+MatchingElementsFillerTest::fill_matching_elements(Query &&query)
+{
+    _matching_elements_filler.reset();
+    _matching_elements.reset();
+    _query = std::move(query);
+    _field_searcher_map.prepare(_index_to_field_ids, _shared_searcher_buf, _query);
+    _matching_elements_filler = std::make_unique<MatchingElementsFiller>(_field_searcher_map, _query, _hit_collector, _search_result);
+    _matching_elements = _matching_elements_filler->fill_matching_elements(_struct_field_mapper);
+}
+
+void
+MatchingElementsFillerTest::assert_elements(uint32_t doc_lid, const vespalib::string& field, const ElementVector& exp_elements)
+{
+    auto act_elements = _matching_elements->get_matching_elements(doc_lid, field);
+    EXPECT_EQ(exp_elements, act_elements);
+}
+
+void
+MatchingElementsFillerTest::assert_same_element(const vespalib::string& field, const vespalib::string& term1, const vespalib::string& term2, const ElementVector& exp_elements)
+{
+    fill_matching_elements(make_same_element(field, term1, term2));
+    assert_elements(1, field, exp_elements);
+}
+
+void
+MatchingElementsFillerTest::assert_same_element_single(const vespalib::string& field, const vespalib::string& term, const ElementVector& exp_elements)
+{
+    fill_matching_elements(make_same_element_single(field + "." + term));
+    assert_elements(1, field, exp_elements);
+}
+
+TEST_F(MatchingElementsFillerTest, same_element_operator)
+{
+    assert_same_element("elem_array", "name:bar", "weight:20", { 1 });
+    assert_same_element("elem_array", "name:zap", "weight:20", { 4, 5 });
+    assert_same_element("elem_map", "value.name:bar", "value.weight:20", { 0 });
+    assert_same_element("elem_map", "value.name:zap", "value.weight:20", { 4, 5 });
+    assert_same_element("str_int_map", "key:bar", "value:20", { 0 });
+    assert_same_element("str_int_map", "key:zap", "value:20", { 4, 5 });
+}
+
+TEST_F(MatchingElementsFillerTest, nested_field)
```

Comment:
Consider being more specific in test name (e.g. matching_elements_calculated_when_searching_on_nested_field)